### PR TITLE
Adds JNumber conversions

### PR DESCRIPTION
**Pull request by eed3si9n**
Ref https://github.com/mdedetrich/scalajson/pull/38

This implementation is loosely based on that of Circe.

`JNumber` is internally subdivided into _package private_ seven datatypes `JStringDecimal`, `JBigDecimal`, `JBigInt`, `JLong`, `JInt`, `JDouble`, and `JFloat` corresponding to the input values passed into `JNumber.apply(...)`. This allows performant roundtrip back to the original number types, if needed.

Note that the parser implementer now has an option of constructing `JStringDecimal` via `JNumber(value: String)` or doing the number parsing upfront.

### CODE REVIEW

File: js/src/main/scala/scalajson/ast/JValue.scala [hunk 1]
```diff
@@ -48,86 +48,281 @@ final case class JString(value: String) extends JValue {
...
+    case jNumberRegex(_ *) => Option(JStringDecimal(value))
+    case _ => None
   }
 
-  def apply(value: String): Option[JNumber] =
-    fromString(value)
+  def unapply(value: JNumber): Option[String] = Option(value.stringValue)
+
+  private[ast] def jnumberEquals(value: JNumber, obj: Any): Boolean =
+    (value, obj) match {
```
Comment (sirthias) [inline js/src/main/scala/scalajson/ast/JValue.scala:151]: Seeing as `jnumberEquals` might be hot in certain kinds of applications: Could this be expressed in an allocation-free manner (possibly sacrificing a bit of conciseness)?

**Comment by sirthias** [comment]
IMHO representing JSON numbers via an ADT like this is the way to go.
What's you rationale for making the sub types package private instead of opening up the ADT altogether?

**Comment by mdedetrich** [comment]
@sirthias I suppose if the internals change then you are not breaking it for end users?

**Comment by sirthias** [comment]
IMHO this whole effort is about not changing anything, ever.
We need to get it almost 100% right from the beginning.

This is not a library that receives new features or other upgrades over time.
It's supposed to deliver a minimal and consistent JSON ADT, not more and not less.

That's why simplicity and clarity is quite important in my view. Any special construct, hidden fields, non-case-classes or other "tricks" should be extremely well justified.

**Comment by mdedetrich** [comment]
> This is not a library that receives new features or other upgrades over time.
It's supposed to deliver a minimal and consistent JSON ADT, not more and not less.

Completely agree, which is why I am being very careful about a release. Will need @eed3si9n , and possibly @travisbrown to clarify why they are private

**Comment by eed3si9n** [comment]
> What's you rationale for making the sub types package private instead of opening up the ADT altogether?

Unlike Json4s's AST where you get either `case class JLong(num: Long)` or `case class JDecimal(num: BigDecimal)` representing the actual number value, we have seven that represents how the `JNumber` was constructed. The reason why it's there is to avoid conversion cost, and it's almost irrelevant to the consumer of the AST.

For example, suppose I am getting "Twitter fave count", and I assume the number can be represented as `Long`, depending on the parser the number might be stored as `JStringDecimal`, `JBigDecimal`, `JLong` or `JInt` etc. The only thing I need to be concerned is can I safely convert it to `Long`.

**Comment by eed3si9n** [comment]
> IMHO this whole effort is about not changing anything, ever.
> We need to get it almost 100% right from the beginning.

I agree with the general idea, but we should relax this to keeping strict binary compatibility of the exposed classes.

Other goals are correctness, user-friendliness, and performance. Hiding often allows us to do performance tricks without compromising on the usability of the surface API.

File: jvm/src/main/scala-2.10/scalajson.ast/JValue.scala [hunk 2]
```diff
@@ -1,212 +0,0 @@
...
-}
-
-/** Represents a JSON number value. If you are passing in a
-  * NaN or Infinity as a [[scala.Double]] or [[scala.Float]], [[JNumber]] will
-  * return a [[JNull]].
-  *
-  * @author Matthew de Detrich
-  */
-// Due to a restriction in Scala 2.10, we cant override/replace the default apply method
-// generated by the compiler even when the constructor itself is marked private
```
Comment (eed3si9n) [inline jvm/src/main/scala-2.10/scalajson.ast/JValue.scala:92]: Since I can compile my JNumber in 2.10, 2.10 duplication is dropped.

**Comment by mdedetrich** [comment]
@eed3si9n Thanks, I will have a look at this today/tomorrow

**Comment by sirthias** [comment]
> Other goals are correctness, user-friendliness, and performance. Hiding often allows us to do performance tricks without compromising on the usability of the surface API.

Yes, I fully agree. However, in my view this project is a bit different from ordinary libraries as it provides almost only structure and very little logic (at least I would consider this a goal).
If there's no logic we don't really need to hide anything. Also, if the structure is right there is no need for "performance tricks".

My concern with hiding the internal ADT structure is that there will definitely be people that will *want* to access the underlying inner structure, if only to perf-optimize certain cases on their layer.

If we come up with right structure anyone who doesn't care about details can simply rely on the `JNumber` interface. But if I really wanted to I could drop down to the fully specialized `JNumber` ADT and apply dedicated handling to each case.
(One example that come to mind would be an optics library that provides AST transformation/manipulation facilities from the outside. Not being able to see the real underlying structure could be a disadvantage.)

**Comment by eed3si9n** [comment]
@sirthias After talking with @Ichoran on Gitter, I am coming to the idea that both #38 and this PR have feature creeped.

I think you're right that we should try to keep the AST open, and in our case plain `String` (with regex check).

**Comment by mdedetrich** [comment]
Regarding #38 , this was an attempt to provide performant JNumber conversions without altering the current API design too much, hence why the JNumber constructor was largely unchanged and we are using a bitflag

**Comment by eed3si9n** [comment]
Closing this in favor of #42

**Pull request closed without merging**